## Dashboard 2: visualización y documentación para narrar a un equipo de trabajo

- Jesus Alberto Parada Perez

In [1]:
import pandas as pd

### Modelar los datos  

Para poder visualizar y documentar tienen que realizar los siguientes pasos de modelado primero: 

1. Extraer la base de datos “Ask A Manager Salary Survey 2021” de AskAManager.org (https://docs.google.com/spreadsheets/d/1IPS5dBSGtwYVbjsfbaMCYIWnOuRmJcbequohNxCyGVw/edit?resourcekey#gid=1625408792)

In [2]:
## Cargar la base de datos
BD_original = pd.read_csv('Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.csv')
# BD_original.head()
## Nombres de las columnas y dimensiones de la BD
print(f'La base de datos “Ask A Manager Salary Survey 2021” tiene {BD_original.shape[0]} observaciones/respuestas y \
{BD_original.shape[1]} variables/columnas. \nA continuación se presentan los nombres de las variables:\n')

def variables_tipos(df):
    for col in df.columns:
        if str(df[col].dtypes) == 'object':
            print(f'{col} : {df[col].dtypes} (texto).')
        else:
            print(f'{col} : {df[col].dtypes} (numerica).')
            
variables_tipos(BD_original)

La base de datos “Ask A Manager Salary Survey 2021” tiene 27935 observaciones/respuestas y 18 variables/columnas. 
A continuación se presentan los nombres de las variables:

Timestamp : object (texto).
How old are you? : object (texto).
What industry do you work in? : object (texto).
Job title : object (texto).
If your job title needs additional context, please clarify here: : object (texto).
What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.) : object (texto).
How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits. : float64 (numerica).
Please indicate the currency : object (texto).
If "Other," please indicate the currency here:  : object (texto).
If your income needs additional

Una primera exploración de los datos cargados nos permite saber que nuestra base de datos contiene información de 27935 personas distribuida en 17 variables y la fecha de diligenciamiento de la encuesta, con preguntas como: edad, sector de trabajo, puesto de trabajo, información adicional del puesto de trabajo, salario anual, compensaciones adicionales a salario, moneda, otras monedas, información adicional de los ingresos, pais donde trabaja, estado donde trabaja (solo USA), ciudad donde trabaja, años de experiencia profesional, años de experiencia en el campo de trabajo, nivel educativo (el mas alto completado), genero y raza.

In [3]:
# Diccionario con los nombres de las variables (old) y los nuevos nombres
# list(BD_original.columns)
names_old_new = {'Timestamp' : 'Fecha',
                 'How old are you?' : 'Edad',
                 'What industry do you work in?' : 'Industria',
                 'Job title' : 'Cargo',
                 'If your job title needs additional context, please clarify here:' : 'Contexto cargo',
                 "What is your annual salary? (You'll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)" : 'Salario anual',
                 'How much additional monetary compensation do you get, if any (for example, bonuses or overtime in an average year)? Please only include monetary compensation here, not the value of benefits.' : 'Otras compensaciones', 
                 'Please indicate the currency' : 'Moneda',
                 'If "Other," please indicate the currency here: ' : 'Otra moneda',
                 'If your income needs additional context, please provide it here:' : 'Contexto salario',
                 'What country do you work in?' : 'Pais',
                 "If you're in the U.S., what state do you work in?" : 'Estado (USA)',
                 'What city do you work in?' : 'Ciudad',
                 'How many years of professional work experience do you have overall?' : 'experiencia',
                 'How many years of professional work experience do you have in your field?' : 'experiencia especifica',
                 'What is your highest level of education completed?' : 'Educacion',
                 'What is your gender?' : 'Genero',
                 'What is your race? (Choose all that apply.)' : 'Raza'
                }
BD_nuevo = BD_original.rename(columns = names_old_new, inplace=False)
BD_nuevo.head()

Fecha   Edad                      Industria  \
0  4/27/2021 11:02:10  25-34   Education (Higher Education)   
1  4/27/2021 11:02:22  25-34              Computing or Tech   
2  4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3  4/27/2021 11:02:41  25-34                     Nonprofits   
4  4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   

                                      Cargo Contexto cargo Salario anual  \
0        Research and Instruction Librarian            NaN        55,000   
1  Change & Internal Communications Manager            NaN        54,600   
2                      Marketing Specialist            NaN        34,000   
3                           Program Manager            NaN        62,000   
4                        Accounting Manager            NaN        60,000   

   Otras compensaciones Moneda Otra moneda Contexto salario            Pais  \
0                   0.0    USD         NaN              NaN   United States   
1                4000.0    GBP         NaN              NaN  United Kingdom   
2                   NaN    USD         NaN              NaN              US   
3                3000.0    USD         NaN              NaN             USA   
4                7000.0    USD         NaN              NaN              US   

     Estado (USA)       Ciudad   experiencia experiencia especifica  \
0   Massachusetts       Boston     5-7 years              5-7 years   
1             NaN    Cambridge  8 - 10 years              5-7 years   
2       Tennessee  Chattanooga   2 - 4 years            2 - 4 years   
3       Wisconsin    Milwaukee  8 - 10 years              5-7 years   
4  South Carolina   Greenville  8 - 10 years              5-7 years   

         Educacion      Genero   Raza  
0  Master's degree       Woman  White  
1   College degree  Non-binary  White  
2   College degree       Woman  White  
3   College degree       Woman  White  
4   College degree       Woman  White

En este punto los nombres de las variables fueron reemplazados por nombres mas cortos a través de un diccionario.

2. Limpiar los campos de “Country” y “City” para homogenizar los nombres de los lugares.

In [4]:
## Limpiar Country : Pais
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais'].str.strip() # Eliminar espacios en blanco al principio y al final de la cadena
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.lower() #minusculas 
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].astype(str).apply(lambda x: x.replace('.','')) #eliminar puntos
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('united arab emirates','UAE') #reemplazar por UAE
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('uae','UAE') #reemplazar por UAE
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('brazil','brasil') #reemplazar por brasil
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('california','Estados Unidos') #reemplazar por USA

BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].replace(regex={r'^u.*s$':'Estados Unidos', #reemplazar por USA
                                                               r'^australi.*$':'Australia', #reemplazar por Australia
                                                               r'^austri.*$':'Austria', #reemplazar por Austria
                                                               r'^.*zealand.*$':'New Zealand', #reemplazar por New Zealand
                                                               r'^can.*$':'Canada', #reemplazar por Canada
                                                               r'^.*germany.*$':'Alemania', #reemplazar por Alemania
                                                               r'^d.*mark$':'Dinamarca', #reemplazar por Dinamarca
                                                               r'^ukra.*$':'Ucrania', #reemplazar por Ukrania
                                                               r'^uk.*$':'United Kingdom', #reemplazar por United Kingdom
                                                               r'^uni.*kin.*$':'United Kingdom', #reemplazar por United Kingdom
                                                               r'^engl.*$':'United Kingdom', #reemplazar por United Kingdom
                                                               r'^can.*$':'Canada', #reemplazar por Canada
                                                               r'^wales.*$':'Wales', #reemplazar por Wales
                                                               r'^argenti.*$':'Argentina'}) #reemplazar por Argentina

BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('csnada','Canada') #reemplazar por Canada
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('nz','New Zealand') #reemplazar por New Zealand
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('méxico','mexico') #reemplazar por mexico
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace('san francisco','Estados Unidos') #reemplazar por USA
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.replace("i work for an us based company but i'm from argentina",
                                                            'Argentina') #reemplazar por Argentina

BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].replace({'united y':'Estados Unidos',
                                                         'unitef stated':'Estados Unidos',
                                                         'uniter statez':'Estados Unidos',
                                                         'uxz':'Estados Unidos',
                                                         'ua':'Estados Unidos',
                                                         'usd':'Estados Unidos',
                                                         'virginia':'Estados Unidos',
                                                         'us>':'Estados Unidos',
                                                         'the us':'Estados Unidos',
                                                         'the united states':'Estados Unidos',
                                                         'is':'Otro',
                                                         'isa':'Otro',
                                                         '🇺🇸':'Estados Unidos'}) #reemplazar por USA

BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].replace(regex={r'^italy.*$':'Italia', #reemplazar por Italia
                                                               r'^japan.*$':'Japon', #reemplazar por Japon
                                                               r'^luxemb.*$':'Luxemburgo', #reemplazar por Luxemburgo
                                                               r'^.*ne.*erland.*$':'Holanda', #reemplazar por Holanda
                                                               r'^.*ireland.*$':'Irlanda', #reemplazar por Irlanda
                                                               r'^.*philippines.*$':'philippines', #reemplazar por Filipinas
                                                               r'^scotland.*$':'Escocia', #reemplazar por Escocia
                                                               r'^united s.*$':'Estados Unidos', #reemplazar por USA
                                                               r'^hong.*$':'Hong Kong', #reemplazar por Hong Kong
                                                               r'^us.*$':'Estados Unidos', #reemplazar por USA
                                                               r'^i .*us.*$':'Estados Unidos', #reemplazar por USA
                                                               r'^us.*$':'Estados Unidos', #reemplazar por USA
                                                               r'^wo.*us.*$':'Estados Unidos', #reemplazar por USA
                                                               r'^.*romania.*$':'romania', #reemplazar por Romania
                                                               r'^i .*$':'Otro', #reemplazar por Otro
                                                               r'^.*deducted.*$':'Otro', #reemplazar por Otro
                                                               r'^bonus.*$':'Otro', #reemplazar por Otro
                                                               r'^we .*$':'Otro', #reemplazar por Otro
                                                               r'^n/a .*$':'Otro', #reemplazar por Otro
                                                               r'^na.*$':'Otro', #reemplazar por Otro
                                                               r'^y.*$':'Otro', #reemplazar por Otro
                                                               r'^nl.*$':'Otro', #reemplazar por Otro
                                                               r'^otr.*$':'Otro', #reemplazar por Otro
                                                               r'^for the united.*$':'Estados Unidos',
                                                               r'^usa.*$':'Estados Unidos'}) #reemplazar por USA

# BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.capitalize() #mayuscula primera letra
BD_nuevo['Pais_nuevo'] = BD_nuevo['Pais_nuevo'].str.title() #mayuscula la primera letra de cada palabra

len(BD_nuevo['Pais'].unique()), len(BD_nuevo['Pais_nuevo'].unique()), sorted(BD_nuevo['Pais_nuevo'].unique())

(368,
 112,
 ['Afghanistan',
  'Africa',
  'Alemania',
  'America',
  'Argentina',
  'Australia',
  'Austria',
  'Bangladesh',
  'Belgium',
  'Bermuda',
  'Bosnia And Herzegovina',
  'Brasil',
  'Britain',
  'Bulgaria',
  'Cambodia',
  'Canada',
  'Catalonia',
  'Cayman Islands',
  'Chile',
  'China',
  'Colombia',
  'Congo',
  'Contracts',
  'Costa Rica',
  "Cote D'Ivoire",
  'Croatia',
  'Cuba',
  'Currently Finance',
  'Cyprus',
  'Czech Republic',
  'Czechia',
  'Dinamarca',
  'Ecuador',
  'Eritrea',
  'Escocia',
  'Estados Unidos',
  'Estonia',
  'Europe',
  'Finland',
  'France',
  'Ghana',
  'Global',
  'Great Britain',
  'Greece',
  'Hartford',
  'Holanda',
  'Hong Kong',
  'Hungary',
  'India',
  'Indonesia',
  'International',
  'Irlanda',
  'Isle Of Man',
  'Israel',
  'Italia',
  'Jamaica',
  'Japon',
  'Jersey, Channel Islands',
  'Jordan',
  'Kenya',
  'Kuwait',
  'Latvia',
  'Liechtenstein',
  'Lithuania',
  'Luxemburgo',
  'Mainland China',
  'Malaysia',
  'Malta',
  'M

In [5]:
## Limpiar City : Ciudad
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad'].str.strip() # Eliminar espacios en blanco al principio y al final de la cadena
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.lower() #minusculas 
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].astype(str).apply(lambda x: x.replace('.','')) #eliminar puntos
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].astype(str).apply(lambda x: x.replace('<','')) #eliminar puntos

BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('allendale, mi','Allendale') #reemplazar por Allendale
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('allentown','Allentown') #reemplazar por Allentown
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('bahía blanca, buenos aires, province argentina',
                                                                'buenos aires') #reemplazar por buenos aires
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('city of buenos aires', 'buenos aires') #reemplazar por buenos aires
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('barxelona','barcelona') #reemplazar por barcelona
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('buffalo, ny','buffalo') #reemplazar por buffalo
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('charlottesville','Charlottesville') #reemplazar por Charlottesville
BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].str.replace('charlottetown','Charlottetown') #reemplazar por Charlottetown


BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].replace(regex={r'^aberdeen.*$':'Aberdeen', #reemplazar por Aberdeen
                                                                   r'^abington.*$':'Abington', #reemplazar por Abington
                                                                   r'^addis.*$':'Addison', #reemplazar por Addison
                                                                   r'^adelaide.*$':'Adelaide', #reemplazar por Adelaide
                                                                   r'^akron.*$':'Akron', #reemplazar por Akron
                                                                   r'^albany.*$':'Albany', #reemplazar por Albany
                                                                   r'^albuquerq.*$':'Albuquerque', #reemplazar por Albuquerque
                                                                   r'^aldie.*$':'Aldie', #reemplazar por Aldie
                                                                   r'^alexandri.*$':'Alexandria', #reemplazar por Alexandria
                                                                   r'^allen.*$':'Allen', #reemplazar por Allen
                                                                   r'^alpharetta.*$':'Alpharetta', #reemplazar por Alpharetta
                                                                   r'^amherst.*$':'Amherst', #reemplazar por Amherst
                                                                   r'^amsterd.*$':'Amsterdam', #reemplazar por Amsterdam
                                                                   r'^anaheim.*$':'Anaheim', #reemplazar por Anaheim
                                                                   r'^anchorag.*$':'Anchorage', #reemplazar por Anchorage
                                                                   r'^ann arbor.*$':'Ann arbor', #reemplazar por Ann arbor
                                                                   r'^annandale.*$':'Annandale', #reemplazar por Annandale
                                                                   r'^annapolis.*$':'Annapolis', #reemplazar por Annapolis
                                                                   r'^antwerp.*$':'Antwerpen', #reemplazar por Antwerpen
                                                                   r'^arlington.*$':'Arlington', #reemplazar por Arlington
                                                                   r'^asheville.*$':'Asheville', #reemplazar por Asheville
                                                                   r'^atlanta.*$':'Atlanta', #reemplazar por Atlanta
                                                                   r'^augusta.*$':'Augusta', #reemplazar por Augusta
                                                                   r'^austin.*$':'Austin', #reemplazar por Austin
                                                                   r'^baltimore.*$':'Baltimore', #reemplazar por Baltimore
                                                                   r'^bango.*$':'Bangor', #reemplazar por Bangor
                                                                   r'^barrie.*$':'Barrie', #reemplazar por Barrie
                                                                   r'^bay a.*$':'bay city', #reemplazar por Bay city
                                                                   r'^beaverton.*$':'Beaverton', #reemplazar por Beaverton
                                                                   r'^bedford.*$':'Bedford', #reemplazar por Bedford
                                                                   r'^belfast.*$':'Belfast', #reemplazar por Belfast
                                                                   r'^bellevue.*$':'Bellevue', #reemplazar por Bellevue
                                                                   r'^bellingham.*$':'Bellingham', #reemplazar por Bellingham
                                                                   r'^belmont.*$':'Belmont', #reemplazar por Belmont
                                                                   r'^bend.*$':'Bend', #reemplazar por Bend
                                                                   r'^benton.*$':'Benton', #reemplazar por Benton
                                                                   r'^bergen.*$':'Bergen', #reemplazar por Bergen
                                                                   r'^berkeley.*$':'Berkeley', #reemplazar por Berkeley
                                                                   r'^berkley.*$':'Berkley', #reemplazar por Berkley
                                                                   r'^berlin.*$':'Berlin', #reemplazar por Berlin
                                                                   r'^bethesda.*$':'Bethesda', #reemplazar por Bethesda
                                                                   r'^bethlehem.*$':'Bethlehem', #reemplazar por Bethlehem
                                                                   r'^beverly.*$':'Beverly', #reemplazar por Beverly
                                                                   r'^birm.*$':'Birmingham', #reemplazar por Birmingham
                                                                   r'^bloomfield.*$':'Bloomfield', #reemplazar por Bloomfield
                                                                   r'^bloomington.*$':'Bloomington', #reemplazar por Bloomington
                                                                   r'^boston.*$':'Boston', #reemplazar por Boston
                                                                   r'^boulder.*$':'Boulder', #reemplazar por Boulder
                                                                   r'^bozeman.*$':'Bozeman', #reemplazar por Bozeman
                                                                   r'^brea.*$':'Brea', #reemplazar por Brea
                                                                   r'^bridgewater.*$':'Bridgewater', #reemplazar por Bridgewater
                                                                   r'^brighton.*$':'Brighton', #reemplazar por Brighton
                                                                   r'^brisbane.*$':'Brisbane', #reemplazar por Brisbane
                                                                   r'^bristo.*$':'Bristol', #reemplazar por Bristol
                                                                   r'^british columbia.*$':'British columbia', #reemplazar por British columbia
                                                                   r'^brooklyn.*$':'Brooklyn', #reemplazar por Brooklyn
                                                                   r'^brussel.*$':'Brussels', #reemplazar por Brussels
                                                                   r'^bryan.*$':'Bryan', #reemplazar por Bryan
                                                                   r'^burlington.*$':'Burlington', #reemplazar por Burlington
                                                                   r'^burnaby.*$':'Burnaby', #reemplazar por Burnaby
                                                                   r'^calgary.*$':'Calgary', #reemplazar por Calgary
                                                                   r'^camarillo.*$':'Camarillo', #reemplazar por Camarillo
                                                                   r'^cambridge.*$':'Cambridge', #reemplazar por Cambridge
                                                                   r'^campbell.*$':'Campbell', #reemplazar por Campbell
                                                                   r"^can't.*$":'Otro', #reemplazar por Otro
                                                                   r'^canbridge.*$':'Cambridge', #reemplazar por Cambridge
                                                                   r'^carmel.*$':'Carmel', #reemplazar por Carmel
                                                                   r'^carson.*$':'Carson', #reemplazar por Carson
                                                                   r'^cartersville.*$':'Cartersville', #reemplazar por Cartersville
                                                                   r'^cary.*$':'Cary', #reemplazar por Cary
                                                                   r'^champaign.*$':'Champaign', #reemplazar por Champaign
                                                                   r'^chantilly.*$':'Chantilly', #reemplazar por Chantilly
                                                                   r'^charleston.*$':'Charleston', #reemplazar por Charleston
                                                                   r'^charlestown.*$':'Charlestown', #reemplazar por Charlestown
                                                                   r'^charlotte.*$':'Charlotte', #reemplazar por Charlotte
                                                                   r'^cherry.*$':'Cherry hill', #reemplazar por Cherry hill
                                                                   r'^cheshire.*$':'Cheshire', #reemplazar por Cheshire
                                                                   r'^cheyenne.*$':'Cheyenne', #reemplazar por Cheyenne
                                                                   r'^chica.*$':'Chicago', #reemplazar por Chicago
                                                                   r'^chilliwack.*$':'Chilliwack', #reemplazar por Chilliwack
                                                                   r'^cinci.*$':'Cincinnati', #reemplazar por Cincinnati
                                                                   r'^cleveland.*$':'Cleveland', #reemplazar por cleveland
                                                                   r'^clifton.*$':'Clifton', #reemplazar por clifton
                                                                   r'^clinton.*$':'Clinton', #reemplazar por clinton
                                                                   r'^coeur .*$': "Coeur d'alene", #reemplazar por Coeur d'alene
                                                                   r'^columbus.*$':'Columbus', #reemplazar por columbus
                                                                   r'^concord.*$':'Concord', #reemplazar por concord
                                                                   r'^cortland.*$':'Cortland', #reemplazar por cortland
                                                                   r'^coventry.*$':'Coventry', #reemplazar por coventry
                                                                   r'^crewe.*$':'Crewe', #reemplazar por crewe
                                                                   r'^cupertino.*$':'Cupertino', #reemplazar por cupertino
                                                                   r'^dallas.*$':'Dallas', #reemplazar por dallas
                                                                   r'^danbury.*$':'Danbury', #reemplazar por danbury
                                                                   r'^danvers.*$':'Danvers', #reemplazar por danvers
                                                                   r'^davenport.*$':'Davenport', #reemplazar por davenport
                                                                   r'^daytona.*$':'Daytona', #reemplazar por daytona
                                                                   r'^dc.*$':'Washington D. C.', #reemplazar por Washington D. C.
                                                                   r'^dearborn.*$':'Dearborn', #reemplazar por dearborn
                                                                   r'^decline.*$':'Otro', #reemplazar por Otro
                                                                   r'^deerfield.*$':'Deerfield', #reemplazar por deerfield
                                                                   r'^delaware.*$':'Delaware', #reemplazar por delaware
                                                                   r'^denver.*$':'Denver', #reemplazar por denver
                                                                   r'^derby.*$':'Derby', #reemplazar por derby
                                                                   r'^des moines.*$':'Des moines', #reemplazar por des moines
                                                                   r'^detroit.*$':'Detroit', #reemplazar por detroit
                                                                   r'^dfw.*$':'Dfw', #reemplazar por dfw
                                                                   r'^dmv metro area.*$':'Washington D. C.', #reemplazar por Washington D. C.
                                                                   r'^.*columbia.*$':'Columbia', #reemplazar por columbia
                                                                   r'^don.*want.*$':'Otro', #reemplazar por Otro
                                                                   r'^dubli.*$':'Dublin', #reemplazar por dubli
                                                                   r'^durham.*$':'Durham', #reemplazar por durham
                                                                   r'^eagan.*$':'Eagan', #reemplazar por eagan
                                                                   r'^eagle.*$':'Eagle', #reemplazar por eagle
                                                                   r'^eau claire.*$':'Eau claire', #reemplazar por eau claire
                                                                   r'^eden prairie.*$':'Eden prairie', #reemplazar por eden prairie
                                                                   r'^edina.*$':'Edina', #reemplazar por edina
                                                                   r'^edinburgh.*$':'Edinburgh', #reemplazar por edinburgh
                                                                   r'^edison.*$':'Edison', #reemplazar por edison
                                                                   r'^edmonton.*$':'Edmonton', #reemplazar por edmonton
                                                                   r'^egg harbor.*$':'Egg harbor', #reemplazar por egg harbor
                                                                   r'^elgin.*$':'Elgin', #reemplazar por elgin
                                                                   r'^elkton.*$':'Elkton', #reemplazar por elkton
                                                                   r'^[0-9].*$':'Otro',
                                                                   r'^[(].*$':'Otro',
                                                                   r'^[/].*$':'Otro',
                                                                   r'^[-].*$':'Otro',
                                                                   r'^[*].*$':'Otro',
                                                                   r'^a .*$':'Otro',
                                                                   r'^.*remote.*$':'Otro',
                                                                   r'^elmhurst.*$':'Elmhurst'}) #reemplazar por Elmhurst


## No se termino el proceso de arreglo de la variable city por tiempo, 
## pero se podria continuar analizando todas las opciones restantes

# lista = ["chicago"]

# for a in lista:
#     BD_nuevo['Ciudad_nuevo'] = BD_nuevo['Ciudad_nuevo'].replace(regex={r'^.`{a}`.*$':'`{a}`'})
    
len(BD_nuevo['Ciudad'].unique()), len(BD_nuevo['Ciudad_nuevo'].unique()), sorted(BD_nuevo['Ciudad_nuevo'].unique())

(4798,
 3408,
 ['',
  'Aberdeen',
  'Abington',
  'Addison',
  'Adelaide',
  'Akron',
  'Albany',
  'Albuquerque',
  'Aldie',
  'Alexandria',
  'Allen',
  'Allendale',
  'Allentown',
  'Alpharetta',
  'Amherst',
  'Amsterdam',
  'Anaheim',
  'Anchorage',
  'Ann arbor',
  'Annandale',
  'Annapolis',
  'Antwerpen',
  'Arlington',
  'Asheville',
  'Atlanta',
  'Augusta',
  'Austin',
  'Baltimore',
  'Bangor',
  'Barrie',
  'Beaverton',
  'Bedford',
  'Belfast',
  'Bellevue',
  'Bellingham',
  'Belmont',
  'Bend',
  'Benton',
  'Bergen',
  'Berkeley',
  'Berkley',
  'Berlin',
  'Bethesda',
  'Bethlehem',
  'Beverly',
  'Birmingham',
  'Bloomfield',
  'Bloomington',
  'Boston',
  'Boulder',
  'Bozeman',
  'Brea',
  'Bridgewater',
  'Brighton',
  'Brisbane',
  'Bristol',
  'Brooklyn',
  'Brussels',
  'Bryan',
  'Burlington',
  'Burnaby',
  'Calgary',
  'Camarillo',
  'Cambridge',
  'Campbell',
  'Carmel',
  'Carson',
  'Cartersville',
  'Cary',
  'Champaign',
  'Chantilly',
  'Charleston',
 

3. Crear 2 campos nuevos: “salario_anual” y “compensaciones” convirtiendo sueldos y compensaciones a Pesos Colombianos basados en la tasa de cambio del día que hacen el ejercicio.

In [6]:
import numpy as np
BD_nuevo['Moneda_nueva'] = np.where(BD_nuevo['Moneda'] == 'Other', BD_nuevo['Otra moneda'], BD_nuevo['Moneda'])

BD_nuevo['Moneda_nueva'] = BD_nuevo['Moneda_nueva'].str.strip() # Eliminar espacios en blanco principio y al final cadena
BD_nuevo['Moneda_nueva'] = BD_nuevo['Moneda_nueva'].str.upper() #mayuscula 
BD_nuevo['Moneda_nueva'] = BD_nuevo['Moneda_nueva'].astype(str).apply(lambda x: x.replace('.','')) #eliminar puntos
BD_nuevo['Moneda_nueva'] = BD_nuevo['Moneda_nueva'].astype(str).apply(lambda x: x.replace('<','')) #eliminar puntos

BD_nuevo['Moneda_nueva'] = BD_nuevo['Moneda_nueva'].replace(regex={r'^AME.*$':'USD', #reemplazar por USD
                                                                   r'^US.*$':'USD', #reemplazar por USD
                                                                   r'^ARGE.*$':'ARS', #reemplazar por ARS
                                                                   r'^ARS.*$':'ARS', #reemplazar por ARS
                                                                   r'^PESO.*$':'ARS', #reemplazar por ARS
                                                                   r'^BR.*$':'BRL', #reemplazar por BRL
                                                                   r'^CROAT.*$':'HRK', #reemplazar por HRK
                                                                   r'^CZECH.*$':'CZK', #reemplazar por CZK
                                                                   r'^DANISH.*$':'DKK', #reemplazar por DKK
                                                                   r'^EUR.*$':'EUR', #reemplazar por EUR
                                                                   r'^AU.*$':'AUD/NZD', #reemplazar por AUD/NZD
                                                                   r'^CHINA.*$':'CNY', #reemplazar por CNY
                                                                   r'^RM.*$':'CNY', #reemplazar por CNY
                                                                   r'^RUP.*$':'INR', #reemplazar por INR
                                                                   r'^K.*$':'KRW', #reemplazar por KRW
                                                                   r'^MEX.*$':'MXN', #reemplazar por MXN
                                                                   r'^ILS.*$':'ILS', #reemplazar por ILS
                                                                   r'^ISR.*$':'ILS', #reemplazar por ILS
                                                                   r'^NIS.*$':'ILS', #reemplazar por ILS
                                                                   r'^NOR.*$':'NOK', #reemplazar por NOK
                                                                   r'^NTD.*$':'NZD', #reemplazar por NZD
                                                                   r'^PH.*$':'PHP', #reemplazar por PHP
                                                                   r'^PLN.*$':'PLN', #reemplazar por PLN
                                                                   r'^POL.*$':'PLN', #reemplazar por PLN
                                                                   r'^SIN.*$':'SGD', #reemplazar por SGD
                                                                   r'^TAI.*$':'TWD', #reemplazar por TWD
                                                                   r'^THAI.*$':'THB', #reemplazar por THB
                                                                   r'^IN.*$':'INR'}) #reemplazar por INR
                                                           
                                                                   
len(BD_nuevo['Moneda_nueva'].unique()), list(sorted(BD_nuevo['Moneda_nueva'].unique()))
# BD_nuevo

(38,
 ['ARS',
  'AUD/NZD',
  'BDT',
  'BRL',
  'CAD',
  'CHF',
  'CNY',
  'COP',
  'CZK',
  'DKK',
  'EQUITY',
  'EUR',
  'GBP',
  'HKD',
  'HRK',
  'IDR',
  'ILS',
  'INR',
  'JPY',
  'KRW',
  'LKR',
  'MXN',
  'MYR',
  'NGN',
  'NOK',
  'NZD',
  'PHP',
  'PLN',
  'SAR',
  'SEK',
  'SGD',
  'THB',
  'TRY',
  'TTD',
  'TWD',
  'USD',
  'ZAR',
  'nan'])

In [7]:
# Crear salario y compensaciones en pesos colombianos
exchange_rate = {'ARS':25.089164,'AUD/NZD':3302.8366,'BDT':44.759419,'BRL':916.5687,'CAD':3544.5835,'CHF':5180.1792, 
                 'CNY':701.60188,'COP':1,'CZK':214.98247,'DKK':686.48991,'EQUITY':1,'EUR':5111.0839,
                 'GBP':5754.4869,'HKD':606.90176,'HRK':677.81578,'IDR':0.31472073,'ILS':1363.7631,'INR':57.619925,   
                 'JPY':36.300941,'KRW':3.7763487,'LKR':13.036674,'MXN':251.47472,'MYR':1107.6166,'NGN':10.345028,   
                 'NOK':461.77983,'NZD':3004.6729,'PHP':86.886681,'PLN':1077.6163,'SAR':1270.0934,'SEK':450.30553,
                 'SGD':3592.0935,'THB':142.22123,'TRY':252.93638,'TTD':701.2711,'TWD':158.31982,'USD':4763.9159,
                 'ZAR':268.30007,'nan':1}

# df['Compensation in COP'] = df.apply(lambda x: x['Compensation']*tasas_cambio_dia_ejercicio[x['Currency']], axis=1)

BD_nuevo['Tasa_cambio'] = BD_nuevo['Moneda_nueva'].map(exchange_rate)
BD_nuevo['Salario anual'] = BD_nuevo['Salario anual'].astype(str).apply(lambda x: x.replace(',','')) #eliminar puntos
BD_nuevo['Salario anual'] = BD_nuevo['Salario anual'].astype(float)
BD_nuevo['Salario anual'] = BD_nuevo['Salario anual'].fillna(0) #reemplazar missing con ceros
BD_nuevo['Salario_COP'] = BD_nuevo['Salario anual']*BD_nuevo['Tasa_cambio']
BD_nuevo['Otras compensaciones'] = BD_nuevo['Otras compensaciones'].fillna(0) #reemplazar missing con ceros
BD_nuevo['Compensaciones_COP'] = BD_nuevo['Otras compensaciones']*BD_nuevo['Tasa_cambio']
            

pd.set_option('display.float_format', lambda x: '%.1f' % x)        
BD_nuevo


Fecha   Edad                      Industria  \
0      4/27/2021 11:02:10  25-34   Education (Higher Education)   
1      4/27/2021 11:02:22  25-34              Computing or Tech   
2      4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3      4/27/2021 11:02:41  25-34                     Nonprofits   
4      4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   
...                   ...    ...                            ...   
27930  1/31/2023 17:52:38  25-34    Marketing, Advertising & PR   
27931  1/31/2023 18:13:07  35-44                     Nonprofits   
27932    2/1/2023 7:53:38  45-54                            Law   
27933   2/2/2023 14:22:11  18-24              Computing or Tech   
27934   2/5/2023 23:44:57  18-24   Education (Higher Education)   

                                          Cargo  \
0            Research and Instruction Librarian   
1      Change & Internal Communications Manager   
2                          Marketing Specialist   
3                               Program Manager   
4                            Accounting Manager   
...                                         ...   
27930                            Brand manager    
27931                              label editor   
27932                                  Attorney   
27933                           DevOps Engineer   
27934                                     tutor   

                                          Contexto cargo  Salario anual  \
0                                                    NaN        55000.0   
1                                                    NaN        54600.0   
2                                                    NaN        34000.0   
3                                                    NaN        62000.0   
4                                                    NaN        60000.0   
...                                                  ...            ...   
27930                                                NaN       120500.0   
27931  I work for a museum and edit the labels that g...        64000.0   
27932  In-house counsel and other duties at a corpora...       110000.0   
27933                                                NaN        90000.0   
27934  I hold several different jobs throughout the y...        27040.0   

       Otras compensaciones Moneda Otra moneda Contexto salario  ...  \
0                       0.0    USD         NaN              NaN  ...   
1                    4000.0    GBP         NaN              NaN  ...   
2                       0.0    USD         NaN              NaN  ...   
3                    3000.0    USD         NaN              NaN  ...   
4                    7000.0    USD         NaN              NaN  ...   
...                     ...    ...         ...              ...  ...   
27930               12000.0    CAD         NaN              NaN  ...   
27931                   0.0    USD         NaN              NaN  ...   
27932               12000.0    USD         NaN              NaN  ...   
27933               10000.0    USD         NaN              NaN  ...   
27934                   0.0    USD         NaN              NaN  ...   

      experiencia especifica                           Educacion      Genero  \
0                  5-7 years                     Master's degree       Woman   
1                  5-7 years                      College degree  Non-binary   
2                2 - 4 years                      College degree       Woman   
3                  5-7 years                      College degree       Woman   
4                  5-7 years                      College degree       Woman   
...                      ...                                 ...         ...   
27930              5-7 years                      College degree       Woman   
27931              5-7 years                     Master's degree       Woman   
27932              5-7 years  Professional degree (MD, JD, etc.)         NaN   
27933            2 - 4 years        

4. Crear un campo adicional sumando salario anual y compensaciones en pesos colombianos.

In [8]:
BD_nuevo['Ingreso_total'] = BD_nuevo['Compensaciones_COP'] + BD_nuevo['Salario_COP']
BD_nuevo


Fecha   Edad                      Industria  \
0      4/27/2021 11:02:10  25-34   Education (Higher Education)   
1      4/27/2021 11:02:22  25-34              Computing or Tech   
2      4/27/2021 11:02:38  25-34  Accounting, Banking & Finance   
3      4/27/2021 11:02:41  25-34                     Nonprofits   
4      4/27/2021 11:02:42  25-34  Accounting, Banking & Finance   
...                   ...    ...                            ...   
27930  1/31/2023 17:52:38  25-34    Marketing, Advertising & PR   
27931  1/31/2023 18:13:07  35-44                     Nonprofits   
27932    2/1/2023 7:53:38  45-54                            Law   
27933   2/2/2023 14:22:11  18-24              Computing or Tech   
27934   2/5/2023 23:44:57  18-24   Education (Higher Education)   

                                          Cargo  \
0            Research and Instruction Librarian   
1      Change & Internal Communications Manager   
2                          Marketing Specialist   
3                               Program Manager   
4                            Accounting Manager   
...                                         ...   
27930                            Brand manager    
27931                              label editor   
27932                                  Attorney   
27933                           DevOps Engineer   
27934                                     tutor   

                                          Contexto cargo  Salario anual  \
0                                                    NaN        55000.0   
1                                                    NaN        54600.0   
2                                                    NaN        34000.0   
3                                                    NaN        62000.0   
4                                                    NaN        60000.0   
...                                                  ...            ...   
27930                                                NaN       120500.0   
27931  I work for a museum and edit the labels that g...        64000.0   
27932  In-house counsel and other duties at a corpora...       110000.0   
27933                                                NaN        90000.0   
27934  I hold several different jobs throughout the y...        27040.0   

       Otras compensaciones Moneda Otra moneda Contexto salario  ...  \
0                       0.0    USD         NaN              NaN  ...   
1                    4000.0    GBP         NaN              NaN  ...   
2                       0.0    USD         NaN              NaN  ...   
3                    3000.0    USD         NaN              NaN  ...   
4                    7000.0    USD         NaN              NaN  ...   
...                     ...    ...         ...              ...  ...   
27930               12000.0    CAD         NaN              NaN  ...   
27931                   0.0    USD         NaN              NaN  ...   
27932               12000.0    USD         NaN              NaN  ...   
27933               10000.0    USD         NaN              NaN  ...   
27934                   0.0    USD         NaN              NaN  ...   

                                Educacion      Genero  \
0                         Master's degree       Woman   
1                          College degree  Non-binary   
2                          College degree       Woman   
3                          College degree       Woman   
4                          College degree       Woman   
...                                   ...         ...   
27930                      College degree       Woman   
27931                     Master's degree       Woman   
27932  Professional degree (MD, JD, etc.)         NaN   
27933                     Master's degree         Man   
27934                        Some college       Woman   

                          Raza      Pais_nuevo Ciudad_nuevo Moneda_nueva  \
0                        White  Estados Unidos       Boston          USD   

In [9]:
## Exportar a csv
BD_nuevo.to_csv('Base_final.csv')